pkg/controller/blockdevice/effects.go

```diff
@@ -0,0 +1,297 @@
+package blockdevice
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	longhornv1 "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhtypes "github.com/longhorn/longhorn-manager/types"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	diskv1 "github.com/harvester/node-disk-manager/pkg/apis/harvesterhci.io/v1beta1"
+	"github.com/harvester/node-disk-manager/pkg/block"
+	"github.com/harvester/node-disk-manager/pkg/disk"
+	ctldiskv1 "github.com/harvester/node-disk-manager/pkg/generated/controllers/harvesterhci.io/v1beta1"
+	ctllonghornv1 "github.com/harvester/node-disk-manager/pkg/generated/controllers/longhorn.io/v1beta1"
+	"github.com/harvester/node-disk-manager/pkg/util"
+)
+
+type effect func(effectController, *diskv1.BlockDevice) error
+
+type effectController interface {
+	Nodes() ctllonghornv1.NodeClient
+	Blockdevices() ctldiskv1.BlockDeviceController
+	BlockdeviceCache() ctldiskv1.BlockDeviceCache
+}
+
+// effectGptPartition makes a new GPT table with a single partitions on given
+// blockdevice, and then turns ProvisionPhase to phasePartitioned.
+func effectGptPartition(e effectController, bd *diskv1.BlockDevice) error {
+	go func() {
+		logEffect(bd).Info("Start partition")
+		cmdErr := disk.MakeGPTPartition(bd.Spec.DevPath)
+		if cmdErr == nil {
+			logEffect(bd).Info("Finish partitioning")
+		} else {
+			logEffect(bd).Errorf("Failed to partition: %v", cmdErr.Error())
+		}
+
+		update := func(bd *diskv1.BlockDevice) *diskv1.BlockDevice {
+			if cmdErr != nil {
+				diskv1.ProvisionPhaseFailed.Set(bd)
+				setDevicePartitionedCondition(bd, corev1.ConditionFalse, cmdErr.Error())
+			} else {
+				// Backwards compatible for LastFormattedAt
+				bd.Status.DeviceStatus.FileSystem.LastFormattedAt = &metav1.Time{Time: time.Now()}
+				diskv1.ProvisionPhasePartitioned.Set(bd)
+				setDevicePartitionedCondition(bd, corev1.ConditionTrue, "")
+			}
+			return bd
+		}
+
+		updateDevice(e, bd, update)
+	}()
+	return nil
+}
+
+// effectEnqueuePrepareFormatPartition enqueues effectPrepareFormatPartition
+func effectEnqueuePrepareFormatPartition(e effectController, bd *diskv1.BlockDevice) error {
+	logEffect(bd).Info("Enqueue for preparing format partition")
+	e.Blockdevices().EnqueueAfter(bd.Namespace, bd.Name, enqueueDelay)
+	return nil
+}
+
+// effectPrepareFormatPartition prepares required spec for child partition before start formating
+func effectPrepareFormatPartition(childBd *diskv1.BlockDevice) effect {
+	return func(e effectController, parentBd *diskv1.BlockDevice) error {
+		logEffect(parentBd).Infof("Start preparing format partition for its child %s", childBd.Name)
+		chilBdCpy := childBd.DeepCopy()
+		chilBdCpy.Spec.FileSystem.MountPoint = parentBd.Spec.FileSystem.MountPoint
```

Comment:
Should the `MountPoint` of the child blockdevice be its name instead of using its parent's `MountPoint`?
```suggestion
		chilBdCpy.Spec.FileSystem.MountPoint = fmt.Sprintf("/var/lib/harvester/extra-disks/%s", chilBdCpy.Name)
```